{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## LOAD FUNCTIONS AND VARIABLES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-01-17T14:47:42.085523Z",
     "iopub.status.busy": "2024-01-17T14:47:42.085192Z",
     "iopub.status.idle": "2024-01-17T14:47:43.381652Z",
     "shell.execute_reply": "2024-01-17T14:47:43.380254Z"
    }
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import math\n",
    "import json\n",
    "import timeit\n",
    "\n",
    "import pandas as pd\n",
    "import seaborn as sns; sns.set()\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import itertools as it\n",
    "\n",
    "from statistics import mean \n",
    "from Bio import SeqIO, Seq\n",
    "from itertools import repeat\n",
    "\n",
    "from matplotlib.colors import LinearSegmentedColormap\n",
    "from matplotlib.gridspec import GridSpec\n",
    "\n",
    "#------------------------------------------\n",
    "cmap1=LinearSegmentedColormap.from_list(\"my_colormap\", sns.color_palette(\"colorblind\", n_colors=5))\n",
    "\n",
    "\n",
    "sns.set_style(\"ticks\",{'axes.grid' : True})\n",
    "sns.set_palette(\"colorblind\")\n",
    "\n",
    "plt.rcParams[\"axes.linewidth\"] = 1.5\n",
    "plt.rcParams[\"xtick.major.width\"] = 1.5\n",
    "plt.rcParams[\"ytick.major.width\"] = 1.5\n",
    "plt.rcParams[\"xtick.major.size\"] = 8\n",
    "plt.rcParams[\"ytick.major.size\"] = 8\n",
    "plt.rcParams[\"axes.titlepad\"] = 20\n",
    "\n",
    "plt.rcParams['svg.fonttype'] = 'none'\n",
    "plt.rcParams[\"axes.titlesize\"] = 30\n",
    "plt.rcParams['axes.labelsize'] = 23.5\n",
    "plt.rcParams['xtick.labelsize'] = 18\n",
    "plt.rcParams['ytick.labelsize'] = 18\n",
    "plt.rcParams['legend.fontsize'] = 18\n",
    "plt.rcParams['font.family'] = 'sans-serif'\n",
    "plt.rcParams['font.sans-serif'] = ['Liberation Sans']\n",
    "plt.rcParams['text.usetex'] = False\n",
    "\n",
    "plt.rcParams['svg.fonttype'] = 'none'\n",
    "plt.rcParams[\"savefig.dpi\"]=300"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-01-17T14:47:43.387002Z",
     "iopub.status.busy": "2024-01-17T14:47:43.386438Z",
     "iopub.status.idle": "2024-01-17T14:47:43.392719Z",
     "shell.execute_reply": "2024-01-17T14:47:43.391895Z"
    }
   },
   "outputs": [],
   "source": [
    "input_genomes=snakemake.params.genome\n",
    "input_sample=snakemake.params.sample\n",
    "input_control=snakemake.params.control\n",
    "input_workdir=snakemake.params.workdir\n",
    "input_figdir=snakemake.params.figdir\n",
    "input_tombodir=snakemake.params.tombo_dir\n",
    "input_threshold=snakemake.params.threshold_modfrac\n",
    "\n",
    "output_modfrac_png=snakemake.output.modfrac_png\n",
    "output_modfrac_kmers=snakemake.output.modfrac_kmers_table\n",
    "output_coverage_png=snakemake.output.coverage_png\n",
    "\n",
    "output_dinucleotide_histogram_pdf=snakemake.output.dinucleotide\n",
    "output_trinucleotide_histogram_pdf=snakemake.output.trinucleotide\n",
    "output_tetranucleotide_histogram_pdf=snakemake.output.tetranucleotide\n",
    "output_pentanucleotide_histogram_pdf=snakemake.output.pentanucleotide\n",
    "# output_hexanucleotide_histogram_pdf=snakemake.output.hexanucleotide\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-01-17T14:47:43.396582Z",
     "iopub.status.busy": "2024-01-17T14:47:43.396295Z",
     "iopub.status.idle": "2024-01-17T14:47:43.404721Z",
     "shell.execute_reply": "2024-01-17T14:47:43.403848Z"
    }
   },
   "outputs": [],
   "source": [
    "threshold=input_threshold\n",
    "genome=input_genomes\n",
    "sample=input_sample\n",
    "control=input_control\n",
    "workdir=input_workdir\n",
    "tombo_dir=input_tombodir + \"/\"\n",
    "\n",
    "%cd $workdir"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-01-17T14:47:43.446176Z",
     "iopub.status.busy": "2024-01-17T14:47:43.444923Z",
     "iopub.status.idle": "2024-01-17T14:47:43.477745Z",
     "shell.execute_reply": "2024-01-17T14:47:43.476740Z"
    }
   },
   "outputs": [],
   "source": [
    "def expand_list(nums):\n",
    "    result = []\n",
    "    for num in nums:\n",
    "        expanded_range = list(range(num - 2, num + 3))  # Creates a list from num-2 to num+2\n",
    "        result.extend(expanded_range)\n",
    "    return sorted(set(result))  # Sorts the list and removes duplicates\n",
    "        \n",
    "def read_genome(fasta_file, strand):\n",
    "    fasta_sequences = SeqIO.parse(open(fasta_file),'fasta')\n",
    "    for fasta in fasta_sequences:\n",
    "        name, sequence = fasta.id, str(fasta.seq)\n",
    "\n",
    "    if strand==\"minus\":\n",
    "        sequence=Seq.reverse_complement(sequence)\n",
    "        \n",
    "    mono=[]\n",
    "    di=[]\n",
    "    tri=[]\n",
    "    tetra=[]\n",
    "    penta=[]\n",
    "    hexa=[]\n",
    "    hepta=[]\n",
    "    octa=[]\n",
    "    nona=[]\n",
    "    deca=[]\n",
    "\n",
    "    lines=sequence + \"X\"*9\n",
    "    for i in range(len(lines)-9):\n",
    "        mono.append(lines[i])\n",
    "        di.append(lines[i]+ lines[i+1])\n",
    "        tri.append(lines[i]+ lines[i+1] + lines[i+2])\n",
    "        tetra.append(lines[i]+ lines[i+1] + lines[i+2] + lines[i+3])\n",
    "        penta.append(lines[i]+ lines[i+1] + lines[i+2] + lines[i+3] + lines[i+4])\n",
    "        hexa.append(lines[i]+ lines[i+1] + lines[i+2] + lines[i+3] + lines[i+4] + lines[i+5])\n",
    "        hepta.append(lines[i]+ lines[i+1] + lines[i+2] + lines[i+3] + lines[i+4] + lines[i+5] + lines[i+6])\n",
    "        octa.append(lines[i]+ lines[i+1] + lines[i+2] + lines[i+3] + lines[i+4] + lines[i+5] + lines[i+6] + lines[i+7])\n",
    "        nona.append(lines[i]+ lines[i+1] + lines[i+2] + lines[i+3] + lines[i+4] + lines[i+5] + lines[i+6] + lines[i+7] + lines[i+8])\n",
    "        deca.append(lines[i]+ lines[i+1] + lines[i+2] + lines[i+3] + lines[i+4] + lines[i+5] + lines[i+6] + lines[i+7] + lines[i+8] + lines[i+9])\n",
    "\n",
    "    d = {'Mononucleotide':mono,'Dinucleotide':di,'Trinucleotide':tri,'Tetranucleotide':tetra,'Pentanucleotide':penta, \"Hexanucleotide\":hexa, \"Heptanucleotide\":hepta, \"Octanucleotide\":octa, \"Nonanucleotide\":nona, \"Decanucleotide\":deca}\n",
    "\n",
    "    nucleotides_df = pd.DataFrame(d)\n",
    "    if strand==\"plus\":\n",
    "        nucleotides_df['Position']=list(range(0,len(lines)-9))\n",
    "    if strand==\"minus\":\n",
    "        nucleotides_df['Position']=list(reversed(range(0,len(lines)-9)))\n",
    "    return nucleotides_df\n",
    "\n",
    "def read_modfrac(wig_file):\n",
    "    greater_thans=[]\n",
    "    sizes=[]\n",
    "    means=[]\n",
    "    position=0\n",
    "    greater_than=0\n",
    "    values=\"\"\n",
    "\n",
    "    n=0\n",
    "    #WIG\n",
    "    with open(wig_file) as fp:\n",
    "        values=[]\n",
    "        positions=[]\n",
    "        greater_than=0\n",
    "        for line in fp:\n",
    "            n=n+1\n",
    "            if n > 2:\n",
    "                values.append(float(line.split()[1]))\n",
    "                positions.append(float(line.split()[0]))\n",
    "                if float(line.split()[1]) > threshold:\n",
    "                    greater_than=greater_than+1\n",
    "    modfrac_df=pd.DataFrame()\n",
    "    modfrac_df[\"position\"]=positions\n",
    "    modfrac_df[\"Mod_fraction\"]=values\n",
    "    return modfrac_df\n",
    "\n",
    "\n",
    "def read_valid_coverage(file_coverage):\n",
    "    #Coverage\n",
    "    coverage_value=[]\n",
    "    coverage_position=[]\n",
    "\n",
    "    #VALID COVERAGE\n",
    "    with open(file_coverage) as fp:\n",
    "        for line in fp:\n",
    "            if not line.startswith(\"track\") and not line.startswith(\"variable\"):\n",
    "                pos, cov=line.split()\n",
    "                cov=int(cov)\n",
    "                pos=int(pos)\n",
    "                coverage_value.append(cov)\n",
    "                coverage_position.append(pos)\n",
    "\n",
    "    coverage_df=pd.DataFrame()\n",
    "    coverage_df[\"position\"]=coverage_position\n",
    "    coverage_df[\"valid_coverage\"]=coverage_value\n",
    "    return coverage_df\n",
    "\n",
    "def read_coverage_wig(wigfile):\n",
    "    coverage=[1]\n",
    "    mean_coverages_sample=[]\n",
    "    genomes_cov_sample=[]\n",
    "    values=\"\"\n",
    "    with open(wigfile) as fp:\n",
    "        for line in fp:\n",
    "            if (not line.startswith(\"track\")) & (not line.startswith(\"variableStep\")) :\n",
    "                genome, low, high, cov=line.split(\"\\t\")\n",
    "                if not genome in genomes_cov_sample:\n",
    "                    genomes_cov_sample.append(genome)\n",
    "                    mean_coverages_sample.append(mean((coverage)))\n",
    "                    coverage=[]\n",
    "                coverage.extend(repeat(int(cov), int(high)-int(low)))\n",
    "    coverage_df=pd.DataFrame()\n",
    "    coverage_df[\"coverage\"]=coverage\n",
    "    coverage_df[\"position\"]=list(range(0,len(coverage)))\n",
    "\n",
    "    return coverage_df\n",
    "\n",
    "def read_modfrac_genome_denovo(genome,barcode, tombo_dir, strand):\n",
    "    fasta_file=\"GENOMES/\" + genome + \".fasta\"\n",
    "\n",
    "      \n",
    "    wig_file= tombo_dir +  genome + \"_\" + barcode + \"_\" + mapping +  \".tombo_denovo/\"+ genome + \"_\" + barcode + \"_denovo.dampened_fraction_modified_reads.\" + strand + \".wig\"\n",
    "    file_valid_coverage= tombo_dir + genome + \"_\" + sample + \"_\" + control + \"_\" + mapping + \".tombo_denovo/\" + genome + \"_\" + sample + \"_\" + control + \"_\" + mapping +  \".valid_coverage.\" + strand + \".wig\"\n",
    "    file_coverage_sample= tombo_dir + genome + \"_\" + sample + \"_\" + control + \"_\" + mapping + \".tombo_denovo/\" + genome + \"_\" + sample + \"_\" + control + \"_\" + mapping +  \".coverage.sample.\" + strand + \".bedgraph\"\n",
    "\n",
    "    nucleotides_df=read_genome(fasta_file, strand)\n",
    "    modfrac_df=read_modfrac(wig_file)\n",
    "    valid_coverage_df=read_valid_coverage(file_valid_coverage)\n",
    "    coverage_sample_df=read_coverage_wig(file_coverage_sample)\n",
    "    coverage_sample_df = coverage_sample_df.rename(columns={'coverage': 'coverage_sample'})\n",
    "\n",
    "    df=nucleotides_df.merge(modfrac_df, left_on=\"Position\", right_on=\"position\", how=\"left\").merge(valid_coverage_df, left_on=\"Position\", right_on=\"position\", how=\"outer\").merge(coverage_sample_df, left_on=\"Position\", right_on=\"position\", how=\"outer\")\n",
    "    df=df.drop(\"position\", axis=1)\n",
    "    df[\"strand\"]=strand\n",
    "    return df\n",
    "\n",
    "def read_modfrac_genome_sampleCompare(genome,sample,control,tombo_dir, strand):\n",
    "    fasta_file=\"GENOMES/\" + genome + \".fasta\"\n",
    "    wig_file= tombo_dir + genome + \"_\" + sample + \"_\" + control + \"_\" + mapping + \".tombo_sampleCompare/\" + genome + \"_\" + sample + \"_\" + control +  \".dampened_fraction_modified_reads.\" + strand + \".wig\"\n",
    "    file_valid_coverage= tombo_dir + genome + \"_\" + sample + \"_\" + control + \"_\" + mapping + \".tombo_sampleCompare/\" + genome + \"_\" + sample + \"_\" + control + \"_\" + mapping +  \".valid_coverage.\" + strand + \".wig\"\n",
    "    file_coverage_sample= tombo_dir + genome + \"_\" + sample + \"_\" + control + \"_\" + mapping + \".tombo_sampleCompare/\" + genome + \"_\" + sample + \"_\" + control + \"_\" + mapping +  \".coverage.sample.\" + strand + \".bedgraph\"\n",
    "    file_coverage_control= tombo_dir + genome + \"_\" + sample + \"_\" + control + \"_\" + mapping + \".tombo_sampleCompare/\" + genome + \"_\" + sample + \"_\" + control + \"_\" + mapping +  \".coverage.control.\" + strand + \".bedgraph\"\n",
    "\n",
    "    nucleotides_df=read_genome(fasta_file, strand)\n",
    "    modfrac_df=read_modfrac(wig_file)\n",
    "    valid_coverage_df=read_valid_coverage(file_valid_coverage)\n",
    "    coverage_sample_df=read_coverage_wig(file_coverage_sample)\n",
    "    coverage_control_df=read_coverage_wig(file_coverage_control)\n",
    "\n",
    "    coverage_sample_df =coverage_sample_df.rename(columns={'coverage': 'coverage_sample'})\n",
    "    coverage_control_df =coverage_control_df.rename(columns={'coverage': 'coverage_control'})\n",
    "\n",
    "    df=nucleotides_df.merge(modfrac_df, left_on=\"Position\", right_on=\"position\", how=\"left\").drop([\"position\"], axis=1).merge(valid_coverage_df, left_on=\"Position\", right_on=\"position\", how=\"outer\").drop([\"position\"], axis=1).merge(coverage_sample_df, left_on=\"Position\", right_on=\"position\", how=\"outer\").drop([\"position\"], axis=1).merge(coverage_control_df, left_on=\"Position\", right_on=\"position\", how=\"outer\")\n",
    "    df=df.drop(\"position\", axis=1)\n",
    "    df[\"strand\"]=strand\n",
    "    return df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## BASIC STATISTICS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-01-17T14:47:43.483135Z",
     "iopub.status.busy": "2024-01-17T14:47:43.482761Z",
     "iopub.status.idle": "2024-01-17T14:47:44.547873Z",
     "shell.execute_reply": "2024-01-17T14:47:44.546603Z"
    }
   },
   "outputs": [],
   "source": [
    "dataframe_sampleCompare_plus=read_modfrac_genome_sampleCompare(genome, sample, control,tombo_dir, \"plus\")\n",
    "dataframe_sampleCompare_minus=read_modfrac_genome_sampleCompare(genome, sample, control,tombo_dir, \"minus\")\n",
    "\n",
    "print()\n",
    "print(\"++++++++++++++++++ PLUS ++++++++++++++++++++\")\n",
    "print()\n",
    "print(\" SAMPLE MEAN COVERAGE + : \\t\",\"%.3f\" % dataframe_sampleCompare_plus[\"coverage_sample\"].mean())\n",
    "print(\"CONTROL MEAN COVERAGE + : \\t\",\"%.3f\" % dataframe_sampleCompare_plus[\"coverage_control\"].mean())\n",
    "print()\n",
    "print(\"  SAMPLE MEAN MODFRAC + : \\t\",\"%.3f\" % dataframe_sampleCompare_plus[\"Mod_fraction\"].mean())\n",
    "print()\n",
    "print()\n",
    "print(\"------------------ MINUS -------------------\")\n",
    "print()\n",
    "print(\" SAMPLE MEAN COVERAGE - : \\t\",\"%.3f\" % dataframe_sampleCompare_minus[\"coverage_sample\"].mean())\n",
    "print(\"CONTROL MEAN COVERAGE - : \\t\",\"%.3f\" % dataframe_sampleCompare_minus[\"coverage_control\"].mean())\n",
    "print()\n",
    "print(\"  SAMPLE MEAN MODFRAC - : \\t\",\"%.3f\" % dataframe_sampleCompare_minus[\"Mod_fraction\"].mean())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MODFRAC PER POSITION (FIRST 10000bp PLUS)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-01-17T14:47:44.553624Z",
     "iopub.status.busy": "2024-01-17T14:47:44.552813Z",
     "iopub.status.idle": "2024-01-17T14:47:56.848864Z",
     "shell.execute_reply": "2024-01-17T14:47:56.847874Z"
    },
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "step=500\n",
    "nrows=math.ceil(len(dataframe_sampleCompare_plus)/step)\n",
    "nrows=min(40,nrows)\n",
    "n=0\n",
    "\n",
    "palette ={\"A\": \"C0\", \"T\": \"C1\", \"G\": \"C2\", \"C\": \"C3\"}\n",
    "fig, axs = plt.subplots(nrows=nrows, ncols=1,figsize=(40,3*nrows), sharey=True)\n",
    "\n",
    "while n < nrows:\n",
    "    min_x=n*step\n",
    "    max_x=(n+1)*step\n",
    "    subset=dataframe_sampleCompare_plus[(dataframe_sampleCompare_plus[\"Position\"]>(min_x)) & (dataframe_sampleCompare_plus[\"Position\"]<max_x)]\n",
    "\n",
    "    filtered_df = subset[subset['Mod_fraction'] > 0.7]\n",
    "    positions_filtered = filtered_df[\"Position\"].to_list()\n",
    "    expanded_positions_filtered = expand_list(positions_filtered)\n",
    "    expanded_filtered_df=subset[subset[\"Position\"].isin(expanded_positions_filtered)]\n",
    "\n",
    "#     ax=sns.barplot(data=subset, x='Position', y='Mod_fraction', ax=axs[n], hue=\"Mononucleotide\", dodge=False, palette=palette)\n",
    "\n",
    "    subset_A=expanded_filtered_df[expanded_filtered_df[\"Mononucleotide\"]==\"A\"]\n",
    "    subset_C=expanded_filtered_df[expanded_filtered_df[\"Mononucleotide\"]==\"C\"]\n",
    "    subset_G=expanded_filtered_df[expanded_filtered_df[\"Mononucleotide\"]==\"G\"]\n",
    "    subset_T=expanded_filtered_df[expanded_filtered_df[\"Mononucleotide\"]==\"T\"]\n",
    "\n",
    "\n",
    "\n",
    "    axs[n].stem(subset_A[\"Position\"], subset_A[\"Mod_fraction\"], 'C0', markerfmt = 'C0o', label = \"A\" )\n",
    "    axs[n].stem(subset_C[\"Position\"], subset_C[\"Mod_fraction\"], 'C3', markerfmt = 'C3o', label = \"C\" )\n",
    "    axs[n].stem(subset_G[\"Position\"], subset_G[\"Mod_fraction\"], 'C2', markerfmt = 'C2o', label = \"G\" )\n",
    "    axs[n].stem(subset_T[\"Position\"], subset_T[\"Mod_fraction\"], 'C1', markerfmt = 'C1o', label = \"T\" )\n",
    "\n",
    "\n",
    "\n",
    "    axs[n].set_ylim(0, 1.1)\n",
    "    axs[n].set_xlim(min_x, max_x)\n",
    "    axs[n].legend(loc=1, fontsize=24)\n",
    "    axs[n].tick_params(axis='x', which='major', labelsize=30)\n",
    "    axs[n].tick_params(axis='y', which='major', labelsize=16)\n",
    "\n",
    "\n",
    "    n=n+1\n",
    "    print(\"{:10.2f}\".format(100*(n)/nrows), \"%\",end = \"\\r\" )\n",
    "\n",
    "\n",
    "plt.tight_layout()\n",
    "fig.savefig(output_modfrac_png)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## COVERAGE PER POSITION "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-01-17T14:47:56.911702Z",
     "iopub.status.busy": "2024-01-17T14:47:56.910105Z",
     "iopub.status.idle": "2024-01-17T14:47:58.232989Z",
     "shell.execute_reply": "2024-01-17T14:47:58.232078Z"
    }
   },
   "outputs": [],
   "source": [
    "step=25000\n",
    "n=0\n",
    "nrows=math.ceil(len(dataframe_sampleCompare_plus)/step)\n",
    "\n",
    "palette ={\"A\": \"C0\", \"T\": \"C1\", \"G\": \"C2\", \"C\": \"C3\"}\n",
    "fig, axs = plt.subplots(nrows=nrows, ncols=1,figsize=(30,8*nrows), sharey=True)\n",
    "\n",
    "while n < nrows:\n",
    "    min_x=n*step\n",
    "    max_x=(n+1)*step\n",
    "    subset_plus=dataframe_sampleCompare_plus[(dataframe_sampleCompare_plus[\"Position\"]>(min_x)) & (dataframe_sampleCompare_plus[\"Position\"]<max_x)]\n",
    "    subset_minus=dataframe_sampleCompare_minus[(dataframe_sampleCompare_minus[\"Position\"]>(min_x)) & (dataframe_sampleCompare_minus[\"Position\"]<max_x)]\n",
    "\n",
    "    sns.lineplot(data=subset_plus, x='Position', y='coverage_sample', ax=axs[n], label=\"sample +\")\n",
    "    sns.lineplot(data=subset_plus, x='Position', y='coverage_control', ax=axs[n], label=\"control +\")\n",
    "    sns.lineplot(data=subset_minus, x='Position', y='coverage_sample', ax=axs[n], label=\"sample -\")\n",
    "    sns.lineplot(data=subset_minus, x='Position', y='coverage_control', ax=axs[n], label=\"control -\")\n",
    "    \n",
    "    axs[n].set_ylabel(\"coverage\")\n",
    "    axs[n].set_xlim(min_x, max_x)\n",
    "\n",
    "    axs[n].legend(loc=1, fontsize=24)\n",
    "    axs[n].tick_params(axis='x', which='major', labelsize=20)\n",
    "\n",
    "\n",
    "    print(\"{:10.2f}\".format(100*n/nrows), \"%\",end = \"\\r\" )\n",
    "    n=n+1\n",
    "\n",
    "\n",
    "plt.tight_layout()\n",
    "fig.savefig(output_coverage_png)\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## KMER STATISTICS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-01-17T14:47:58.243333Z",
     "iopub.status.busy": "2024-01-17T14:47:58.243096Z",
     "iopub.status.idle": "2024-01-17T14:47:58.923689Z",
     "shell.execute_reply": "2024-01-17T14:47:58.922867Z"
    }
   },
   "outputs": [],
   "source": [
    "df=pd.concat([dataframe_sampleCompare_plus,dataframe_sampleCompare_minus]).reset_index()\n",
    "\n",
    "modfrac=df[\"Mod_fraction\"].to_list()\n",
    "modfrac_p1=modfrac[1:] + modfrac[:1]\n",
    "modfrac_p2=modfrac[2:] + modfrac[:2]\n",
    "modfrac_p3=modfrac[3:] + modfrac[:3]\n",
    "modfrac_p4=modfrac[4:] + modfrac[:4]\n",
    "modfrac_p5=modfrac[5:] + modfrac[:5]\n",
    "modfrac_p6=modfrac[6:] + modfrac[:6]\n",
    "modfrac_p7=modfrac[7:] + modfrac[:7]\n",
    "modfrac_p8=modfrac[8:] + modfrac[:8]\n",
    "modfrac_p9=modfrac[9:] + modfrac[:9]\n",
    "modfrac_p10=modfrac[10:] + modfrac[:10]\n",
    "\n",
    "modfrac_m1=modfrac[-1:] + modfrac[:-1]\n",
    "modfrac_m2=modfrac[-2:] + modfrac[:-2]\n",
    "\n",
    "df[\"modfrac_p1\"]=modfrac_p1\n",
    "df[\"modfrac_p2\"]=modfrac_p2\n",
    "df[\"modfrac_p3\"]=modfrac_p3\n",
    "df[\"modfrac_p4\"]=modfrac_p4\n",
    "df[\"modfrac_p5\"]=modfrac_p5\n",
    "df[\"modfrac_p6\"]=modfrac_p6\n",
    "df[\"modfrac_p7\"]=modfrac_p7\n",
    "df[\"modfrac_p8\"]=modfrac_p8\n",
    "df[\"modfrac_p9\"]=modfrac_p9\n",
    "df[\"modfrac_p10\"]=modfrac_p10\n",
    "\n",
    "df[\"modfrac_m1\"]=modfrac_m1\n",
    "df[\"modfrac_m2\"]=modfrac_m2\n",
    "\n",
    "mono=df[\"Mononucleotide\"].to_list()\n",
    "\n",
    "nuc_p1=mono[1:] + mono[:1]\n",
    "nuc_p2=mono[2:] + mono[:2]\n",
    "nuc_p3=mono[3:] + mono[:3]\n",
    "nuc_p4=mono[4:] + mono[:4]\n",
    "nuc_p5=mono[5:] + mono[:5]\n",
    "nuc_p6=mono[6:] + mono[:6]\n",
    "nuc_p7=mono[7:] + mono[:7]\n",
    "nuc_p7=mono[7:] + mono[:7]\n",
    "nuc_p8=mono[8:] + mono[:8]\n",
    "nuc_p9=mono[9:] + mono[:9]\n",
    "nuc_p10=mono[10:] + mono[:10]\n",
    "\n",
    "nuc_m1=mono[-1:] + mono[:-1]\n",
    "nuc_m2=mono[-2:] + mono[:-2]\n",
    "\n",
    "df[\"nuc_p1\"]=nuc_p1\n",
    "df[\"nuc_p2\"]=nuc_p2\n",
    "df[\"nuc_p3\"]=nuc_p3\n",
    "df[\"nuc_p4\"]=nuc_p4\n",
    "df[\"nuc_p5\"]=nuc_p5\n",
    "df[\"nuc_p6\"]=nuc_p6\n",
    "df[\"nuc_p7\"]=nuc_p7\n",
    "df[\"nuc_p7\"]=nuc_p7\n",
    "df[\"nuc_p8\"]=nuc_p8\n",
    "df[\"nuc_p9\"]=nuc_p9\n",
    "df[\"nuc_p10\"]=nuc_p10\n",
    "\n",
    "df[\"nuc_m1\"]=nuc_m1\n",
    "df[\"nuc_m2\"]=nuc_m2\n",
    "df.columns\n",
    "\n",
    "df[['Position','Mononucleotide', 'Dinucleotide', 'Trinucleotide', 'Tetranucleotide',\n",
    "       'Pentanucleotide', 'Hexanucleotide', 'Heptanucleotide', 'Octanucleotide', 'Nonanucleotide', 'Decanucleotide', 'coverage_sample', 'coverage_control', \"modfrac_m2\", \"modfrac_m1\", 'Mod_fraction', 'modfrac_p1', 'modfrac_p2', 'modfrac_p3', 'modfrac_p4',\n",
    "       'modfrac_p5', 'modfrac_p6', 'modfrac_p7', 'modfrac_p8', 'modfrac_p9',\n",
    "       'modfrac_p10']].set_index(\"Position\").to_csv(output_modfrac_kmers)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-01-17T14:47:58.928254Z",
     "iopub.status.busy": "2024-01-17T14:47:58.928044Z",
     "iopub.status.idle": "2024-01-17T14:48:09.670199Z",
     "shell.execute_reply": "2024-01-17T14:48:09.669342Z"
    }
   },
   "outputs": [],
   "source": [
    "categories=[\"Mononucleotide\",\"Dinucleotide\",\"Trinucleotide\",\"Tetranucleotide\",\n",
    "             \"Pentanucleotide\",\"Hexanucleotide\",\"Heptanucleotide\",\n",
    "             \"Octanucleotide\", \"Nonanucleotide\", \"Decanucleotide\"]\n",
    "modfracs=['coverage_sample', 'coverage_control', \"modfrac_m2\",\"modfrac_m1\",\"Mod_fraction\",\"modfrac_p1\", \"modfrac_p2\",\n",
    "          \"modfrac_p3\",\"modfrac_p4\",\"modfrac_p5\",\"modfrac_p6\",\n",
    "          \"modfrac_p7\",\"modfrac_p8\", \"modfrac_p9\",\"modfrac_p10\"]\n",
    "times=[]\n",
    "n=0\n",
    "cmap=\"rainbow\"\n",
    "for category in categories:\n",
    "    start = timeit.default_timer()\n",
    "    sel_cols=modfracs[0:7+n]\n",
    "    sel_cols.append(category)\n",
    "    df_nucleotide=df[sel_cols].groupby(category).mean()\n",
    "    df_count = df[sel_cols].groupby(category).size().rename('count').to_frame()\n",
    "    df_nucleotide[\"max\"]= df_nucleotide[modfracs[2:7+n]].T.max()\n",
    "    df_nucleotide = df_count.join(df_nucleotide)    \n",
    "    df_nucleotide[(~df_nucleotide.index.str.contains(\"X\")) & (df_nucleotide[\"max\"]>0.3)].sort_values(by=\"max\", ascending=False).style.background_gradient(cmap=cmap,vmin=0.15, vmax=1).to_html(input_figdir + \"/\" + genome + \"/sampleCompare_\" + mapping + \"/sampleCompare_\" + genome + \"_\" + sample + \"_\" + control + \"_\" + mapping +\"_table_\"+ category + \".html\")\n",
    "\n",
    "    print( category + \":  \", timeit.default_timer() - start) \n",
    "    times.append(timeit.default_timer() - start)\n",
    "    n=n+1\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## DINUCLEOTIDE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-01-17T14:48:09.675159Z",
     "iopub.status.busy": "2024-01-17T14:48:09.674718Z",
     "iopub.status.idle": "2024-01-17T14:48:12.841251Z",
     "shell.execute_reply": "2024-01-17T14:48:12.840325Z"
    }
   },
   "outputs": [],
   "source": [
    "import itertools as it\n",
    "combinations_temp=list(it.product([\"A\", \"C\", \"G\", \"T\"], repeat=2))\n",
    "combinations=[]\n",
    "for comb in combinations_temp:\n",
    "    combinations.append(''.join(comb))\n",
    "combinations\n",
    "\n",
    "\n",
    "df = df[~df[\"Dinucleotide\"].str.contains(\"X\", na=False)]\n",
    "\n",
    "ncols=4\n",
    "nrows=int(np.ceil(len(combinations)/ncols))\n",
    "fig, axs = plt.subplots(nrows=nrows, ncols=ncols ,figsize=(8,2*nrows), sharey=True)\n",
    "n=0\n",
    "i=0\n",
    "j=0\n",
    "for kmer in combinations:\n",
    "    name=kmer\n",
    "    group=df[df[\"Dinucleotide\"] == kmer]\n",
    "    if len(group)>0:\n",
    "        x=-2\n",
    "        j=int(np.floor((n)/ncols))\n",
    "        i=n-(j*ncols)\n",
    "\n",
    "        for column in [\"modfrac_m2\", \"modfrac_m1\", \"Mod_fraction\", \"modfrac_p1\",  \"modfrac_p2\", \"modfrac_p3\"]:\n",
    "            if x==0 or x==1:\n",
    "                color=palette[name[x]]\n",
    "            else:\n",
    "                color=\"gray\"\n",
    "            data=group[column].to_list()\n",
    "            mean=np.nanmean(data)\n",
    "            std_error = np.nanstd(data, ddof=1)\n",
    "            ax=axs[j,i]\n",
    "            ax.bar(x=x, #x-coordinates of bars\n",
    "               height=mean, #height of bars\n",
    "               yerr=std_error, #error bar width\n",
    "               width=1,\n",
    "               capsize=4, color=color) #length of error bar caps\n",
    "            x=x+1\n",
    "            ax.set_ylim(0, 1)\n",
    "            ax.set_xticks(range(-2, 4, 1))\n",
    "            ax.set_xticklabels([\"N\",\"N\",name[0],name[1],\"N\",\"N\"], )\n",
    "            ax.tick_params(axis='x', which='major', labelsize=16)\n",
    "        ax.set_title(name,x=0.5, y=0.60, fontsize=16)\n",
    "\n",
    "    n=n+1\n",
    "    print(\"{:10.2f}\".format(100*n/len(combinations)), \"%\",end = \"\\r\" )\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.savefig(output_dinucleotide_histogram_pdf, format=\"pdf\", bbox_inches=\"tight\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TRINUCLEOTIDE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-01-17T14:48:12.847182Z",
     "iopub.status.busy": "2024-01-17T14:48:12.846768Z",
     "iopub.status.idle": "2024-01-17T14:48:29.046582Z",
     "shell.execute_reply": "2024-01-17T14:48:29.045715Z"
    }
   },
   "outputs": [],
   "source": [
    "import itertools as it\n",
    "combinations_temp=list(it.product([\"A\", \"C\", \"G\", \"T\"], repeat=3))\n",
    "combinations=[]\n",
    "for comb in combinations_temp:\n",
    "    combinations.append(''.join(comb))\n",
    "combinations\n",
    "\n",
    "df = df[~df[\"Trinucleotide\"].str.contains(\"X\", na=False)]\n",
    "\n",
    "\n",
    "ncols=8\n",
    "nrows=int(np.ceil(len(combinations)/ncols))\n",
    "fig, axs = plt.subplots(nrows=nrows, ncols=ncols ,figsize=(14,2*nrows), sharey=True)\n",
    "n=0\n",
    "i=0\n",
    "j=0\n",
    "for kmer in combinations:\n",
    "    name=kmer\n",
    "    group=df[df[\"Trinucleotide\"] == kmer]\n",
    "    \n",
    "    x=-2\n",
    "    j=int(np.floor((n)/ncols))\n",
    "    i=n-(j*ncols)\n",
    "    \n",
    "    if len(group)>0:\n",
    "\n",
    "        for column in [\"modfrac_m2\", \"modfrac_m1\", \"Mod_fraction\", \"modfrac_p1\",  \"modfrac_p2\", \"modfrac_p3\", \"modfrac_p4\"]:\n",
    "            if x==0 or x==1 or x==2:\n",
    "                color=palette[name[x]]\n",
    "            else:\n",
    "                color=\"gray\"\n",
    "            data=group[column].to_list()\n",
    "            mean=np.nanmean(data)\n",
    "            std_error = np.nanstd(data, ddof=1)\n",
    "            ax=axs[j,i]\n",
    "            ax.bar(x=x, #x-coordinates of bars\n",
    "               height=mean, #height of bars\n",
    "               yerr=std_error, #error bar width\n",
    "               width=.9,\n",
    "               capsize=4, color=color) #length of error bar caps\n",
    "            x=x+1\n",
    "            ax.set_ylim(0, 1)\n",
    "            ax.set_xticks(range(-2, 5, 1))\n",
    "            ax.set_xticklabels([\"N\",\"N\",name[0],name[1],name[2],\"N\",\"N\"], )\n",
    "            ax.tick_params(axis='x', which='major', labelsize=14)\n",
    "            ax.set_title(name,x=0.5, y=0.60, fontsize=18)\n",
    "    \n",
    "    n=n+1\n",
    "    print(\"{:10.2f}\".format(100*n/len(combinations)), \"%\",end = \"\\r\" )\n",
    "plt.tight_layout()\n",
    "plt.savefig(output_trinucleotide_histogram_pdf, format=\"pdf\", bbox_inches=\"tight\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TETRANUCLEOTIDE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-01-17T14:48:29.054106Z",
     "iopub.status.busy": "2024-01-17T14:48:29.053879Z",
     "iopub.status.idle": "2024-01-17T14:49:44.233704Z",
     "shell.execute_reply": "2024-01-17T14:49:44.232537Z"
    }
   },
   "outputs": [],
   "source": [
    "combinations_temp=list(it.product([\"A\", \"C\", \"G\", \"T\"], repeat=4))\n",
    "combinations=[]\n",
    "for comb in combinations_temp:\n",
    "    combinations.append(''.join(comb))\n",
    "combinations\n",
    "\n",
    "\n",
    "df = df[~df[\"Tetranucleotide\"].str.contains(\"X\", na=False)]\n",
    "\n",
    "\n",
    "ncols=16\n",
    "nrows=int(np.ceil(len(combinations)/ncols))\n",
    "fig, axs = plt.subplots(nrows=nrows, ncols=ncols ,figsize=(30,2*nrows), sharey=True)\n",
    "n=0\n",
    "i=0\n",
    "j=0\n",
    "min_modfrac=0.7\n",
    "\n",
    "for kmer in combinations:\n",
    "    name=kmer\n",
    "    group=df[df[\"Tetranucleotide\"] == kmer]\n",
    "    x=-2\n",
    "    j=int(np.floor((n)/ncols))\n",
    "    i=n-(j*ncols)\n",
    "    \n",
    "    if len(group)>0:\n",
    "\n",
    "            #print(i,j)\n",
    "            for column in [\"modfrac_m2\", \"modfrac_m1\", \"Mod_fraction\", \"modfrac_p1\",  \"modfrac_p2\", \"modfrac_p3\", \"modfrac_p4\",\"modfrac_p5\"]:\n",
    "                if x==0 or x==1 or x==2 or x==3:\n",
    "                    color=palette[name[x]]\n",
    "                else:\n",
    "                    color=\"gray\"\n",
    "                data=group[column].to_list()\n",
    "                mean=np.nanmean(data)\n",
    "                std_error = np.nanstd(data, ddof=1)\n",
    "                ax=axs[j,i]\n",
    "                ax.bar(x=x, #x-coordinates of bars\n",
    "                   height=mean, #height of bars\n",
    "                   yerr=std_error, #error bar width\n",
    "                   width=.9,\n",
    "                   capsize=4, color=color) #length of error bar caps\n",
    "                x=x+1\n",
    "            ax.set_ylim(0, 1)\n",
    "            ax.set_xticks(range(-2, 6, 1))\n",
    "            ax.set_xticklabels([\"N\",\"N\",name[0],name[1],name[2],name[3],\"N\",\"N\"], )\n",
    "            ax.tick_params(axis='x', which='major', labelsize=10)\n",
    "            ax.set_title(name,x=0.5, y=0.60, fontsize=18)\n",
    "\n",
    "\n",
    "    n=n+1\n",
    "    print(\"{:10.2f}\".format(100*n/len(combinations)), \"%\",end = \"\\r\" )\n",
    "        \n",
    "plt.tight_layout()\n",
    "plt.savefig(output_tetranucleotide_histogram_pdf, format=\"pdf\", bbox_inches=\"tight\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# PENTANUCLEOTIDE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-01-17T14:49:44.239960Z",
     "iopub.status.busy": "2024-01-17T14:49:44.239727Z",
     "iopub.status.idle": "2024-01-17T14:58:26.476877Z",
     "shell.execute_reply": "2024-01-17T14:58:26.475976Z"
    }
   },
   "outputs": [],
   "source": [
    "combinations_temp=list(it.product([\"A\", \"C\", \"G\", \"T\"], repeat=5))\n",
    "combinations=[]\n",
    "for comb in combinations_temp:\n",
    "    combinations.append(''.join(comb))\n",
    "combinations\n",
    "\n",
    "\n",
    "df = df[~df[\"Pentanucleotide\"].str.contains(\"X\", na=False)]\n",
    "\n",
    "ncols=32\n",
    "nrows=int(np.ceil(len(combinations)/ncols))\n",
    "fig, axs = plt.subplots(nrows=nrows, ncols=ncols ,figsize=(65,2*nrows), sharey=True)\n",
    "n=0\n",
    "i=0\n",
    "j=0\n",
    "\n",
    "min_modfrac=0.7\n",
    "for kmer in combinations:\n",
    "    name=kmer\n",
    "    group=df[df[\"Pentanucleotide\"] == kmer]\n",
    "    x=-2\n",
    "    j=int(np.floor((n)/ncols))\n",
    "    i=n-(j*ncols)\n",
    "    \n",
    "    if len(group)>0:\n",
    "\n",
    "        if ((group[\"Mod_fraction\"].mean() > min_modfrac) or (group[\"modfrac_p1\"].mean() > min_modfrac) or (group[\"modfrac_p2\"].mean() > min_modfrac) or (group[\"modfrac_p3\"].mean() > min_modfrac)):\n",
    "            for column in [\"modfrac_m2\", \"modfrac_m1\", \"Mod_fraction\", \"modfrac_p1\",  \"modfrac_p2\", \"modfrac_p3\", \"modfrac_p4\",\"modfrac_p5\",\"modfrac_p6\"]:\n",
    "                if x==0 or x==1 or x==2 or x==3 or x==4:\n",
    "                    color=palette[name[x]]\n",
    "                else:\n",
    "                    color=\"gray\"\n",
    "                data=group[column].to_list()\n",
    "                mean=np.nanmean(data)\n",
    "                std_error = np.nanstd(data, ddof=1)\n",
    "                ax=axs[j,i]\n",
    "                ax.bar(x=x, #x-coordinates of bars\n",
    "                   height=mean, #height of bars\n",
    "                   yerr=std_error, #error bar width\n",
    "                   width=.9,\n",
    "                   capsize=4, color=color) #length of error bar caps\n",
    "                x=x+1\n",
    "                ax.set_ylim(0, 1)\n",
    "                ax.set_xticks(range(-2, 7, 1))\n",
    "                ax.set_xticklabels([\"N\",\"N\",name[0],name[1],name[2],name[3],name[4],\"N\",\"N\"], )\n",
    "                ax.tick_params(axis='x', which='major', labelsize=14)\n",
    "                ax.set_title(name,x=0.5, y=0.60, fontsize=14)\n",
    "\n",
    "\n",
    "    n=n+1\n",
    "    print(\"{:10.2f}\".format(100*n/len(combinations)), \"%\",end = \"\\r\" )\n",
    "\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.savefig(output_pentanucleotide_histogram_pdf, format=\"pdf\", bbox_inches=\"tight\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## HEXANUCLEOTIDE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2024-01-17T14:58:26.491075Z",
     "iopub.status.busy": "2024-01-17T14:58:26.490843Z",
     "iopub.status.idle": "2024-01-17T14:58:26.496435Z",
     "shell.execute_reply": "2024-01-17T14:58:26.495656Z"
    }
   },
   "outputs": [],
   "source": [
    "# combinations_temp=list(it.product([\"A\", \"C\", \"G\", \"T\"], repeat=6))\n",
    "# combinations=[]\n",
    "# for comb in combinations_temp:\n",
    "#     combinations.append(''.join(comb))\n",
    "# combinations\n",
    "\n",
    "\n",
    "\n",
    "# df = df[~df[\"Hexanucleotide\"].str.contains(\"X\", na=False)]\n",
    "\n",
    "\n",
    "# ncols=64\n",
    "# nrows=int(np.ceil(len(combinations)/ncols))\n",
    "# fig, axs = plt.subplots(nrows=nrows, ncols=ncols ,figsize=(250,2*nrows), sharey=True)\n",
    "# n=0\n",
    "# i=0\n",
    "# j=0\n",
    "# min_modfrac=0.7\n",
    "# for kmer in combinations:\n",
    "#     name=kmer\n",
    "#     group=df[df[\"Hexanucleotide\"] == kmer]\n",
    "#     x=-2\n",
    "#     j=int(np.floor((n)/ncols))\n",
    "#     i=n-(j*ncols)\n",
    "    \n",
    "#     if len(group)>0:\n",
    "\n",
    "#         if ((group[\"Mod_fraction\"].mean() > min_modfrac) or (group[\"modfrac_p1\"].mean() > min_modfrac) or (group[\"modfrac_p2\"].mean() > min_modfrac) or (group[\"modfrac_p3\"].mean() > min_modfrac)):\n",
    "#             for column in [\"modfrac_m2\", \"modfrac_m1\", \"Mod_fraction\", \"modfrac_p1\",  \"modfrac_p2\", \"modfrac_p3\", \"modfrac_p4\",\"modfrac_p5\",\"modfrac_p6\",\"modfrac_p7\"]:\n",
    "#                 if x==0 or x==1 or x==2 or x==3:\n",
    "#                     color=palette[name[x]]\n",
    "#                 else:\n",
    "#                     color=\"gray\"\n",
    "#                 data=group[column].to_list()\n",
    "#                 mean=np.nanmean(data)\n",
    "#                 std_error = np.nanstd(data, ddof=1)\n",
    "#                 ax=axs[j,i]\n",
    "#                 ax.bar(x=x, #x-coordinates of bars\n",
    "#                    height=mean, #height of bars\n",
    "#                    yerr=std_error, #error bar width\n",
    "#                    width=.9,\n",
    "#                    capsize=4, color=color) #length of error bar caps\n",
    "#                 x=x+1\n",
    "#                 ax.set_ylim(0, 1)\n",
    "#                 ax.set_xticks(range(-2, 8, 1))\n",
    "#                 ax.set_xticklabels([\"N\",\"N\",name[0],name[1],name[2],name[3],name[4],name[5],\"N\",\"N\"], )\n",
    "#                 ax.tick_params(axis='x', which='major', labelsize=10)            \n",
    "#                 ax.set_title(name,x=0.5, y=0.60, fontsize=10)\n",
    "#     n=n+1\n",
    "#     print(\"{:10.2f}\".format(100*n/len(combinations)), \"%\",end = \"\\r\" )\n",
    "\n",
    "# plt.tight_layout()\n",
    "# # plt.savefig(output_hexanucleotide_histogram_pdf, format=\"pdf\", bbox_inches=\"tight\")\n",
    "# plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
